### PR TITLE
Indicate shipping costs on product pages.

### LINKED ISSUE #1336: Indicate shipping costs before store checkout.
Copied from a conversion between myself and Dan on G+. 

It'd be nice to have some sort of indication of shipping costs.  Before You click the Checkout button or fill in any information. Ideally on the Product page. (You'll get way more conversions if you do this). 

An indication only. It doesn't have to be exact or cheap. you can use Geo-location features to get an approx location to estimate a cost. Right now you don't know if there even is a cost. 

Letting me calculate shipping or get an indication of it before I commit to filling out all my info in the checkout will go a long way for customers, Esp when international shipping is involved. 

I was pleasantly surprised at $5ish USD shipping to New Zealand. But I couldn't be bothered going through the checkout to find-out if there were shipping costs the first time you shared the updated store. 

Running a web-store can seem really simple, but its extremely easy to sabotage your own success with these little details that can be easily solved.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/37605207-indicate-shipping-costs-before-store-checkout?utm_campaign=plugin&utm_content=tracker%2F10548672&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F10548672&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by lewisgoddard**
Fixes #1336

This pull request is ready for review.

Viewable at https://beta.elementary.io/fix-1336/store/ but doesn't seem to have generated properly.

### Changes Summary

- Fetch approximate location from existing GeoIP
- Add a new API endpoint `geolocate` to estimate shipping address and fees for a given user/product combo.
- Trigger on product modal launch
- Fix close button alignment (icon does not appear on local testing because of FontAwesome licensing)

### API Endpoint
`GET /api/geolocate?shipping&item=6465`

```
{ "ip": "92.26.51.149", "shipping": { "address": { "city": "Colchester", "state": "Essex", "stateCode": "ESS", "country": "United Kingdom", "countryCode": "GB", "postcode": "CO7", "continent": "EU" }, "estimates": [ { "id": "APC_PARCEL_CONNECT", "name": "Int'l Economy without tracking", "expected": "10-20 business days after fulfillment", "cost": "4.04" }, { "id": "APC_PARCEL_CONFIRM", "name": "Int'l Economy with tracking", "expected": "7-16 business days after fulfillment", "cost": "7.00" }, { "id": "USPS_FIRSTCLASSPACKAGEINTERNATIONALSERVICE", "name": "USPS First Class Package Int'l", "expected": "10-20 business days after fulfillment", "cost": "14.11" }, { "id": "APC_PARCEL_EXPEDITED", "name": "Int'l Priority with tracking", "expected": "3-5 business days after fulfillment", "cost": "39.98" }, { "id": "USPS_PRIORITYMAILINTERNATIONAL", "name": "USPS Priority Mail Int'l", "expected": "6-10 business days after fulfillment", "cost": "50.69" }, { "id": "FEDEX_INTERNATIONAL_ECONOMY", "name": "FedEx International Economy", "expected": "4-6 business days after fulfillment", "cost": "57.29" }, { "id": "FEDEX_INTERNATIONAL_PRIORITY", "name": "FedEx International Priority", "expected": "1-3 business days after fulfillment", "cost": "71.37" } ] } }
```

### Before
screenshot from 2018-08-08 12 33 17

### After
screenshot from 2018-08-08 12 37 07

### CODE REVIEW

**Comment by lewisgoddard** [comment]
`GET /api/geolocate?shipping&item=6465`

```
{ "ip": "92.26.51.149", "shipping": { "address": { "city": "Colchester", "state": "Essex", "stateCode": "ESS", "country": "United Kingdom", "countryCode": "GB", "postcode": "CO7", "continent": "EU" }, "estimates": [ { "id": "APC_PARCEL_CONNECT", "name": "Int'l Economy without tracking", "expected": "10-20 business days after fulfillment", "cost": "4.04" }, { "id": "APC_PARCEL_CONFIRM", "name": "Int'l Economy with tracking", "expected": "7-16 business days after fulfillment", "cost": "7.00" }, { "id": "USPS_FIRSTCLASSPACKAGEINTERNATIONALSERVICE", "name": "USPS First Class Package Int'l", "expected": "10-20 business days after fulfillment", "cost": "14.11" }, { "id": "APC_PARCEL_EXPEDITED", "name": "Int'l Priority with tracking", "expected": "3-5 business days after fulfillment", "cost": "39.98" }, { "id": "USPS_PRIORITYMAILINTERNATIONAL", "name": "USPS Priority Mail Int'l", "expected": "6-10 business days after fulfillment", "cost": "50.69" }, { "id": "FEDEX_INTERNATIONAL_ECONOMY", "name": "FedEx International Economy", "expected": "4-6 business days after fulfillment", "cost": "57.29" }, { "id": "FEDEX_INTERNATIONAL_PRIORITY", "name": "FedEx International Priority", "expected": "1-3 business days after fulfillment", "cost": "71.37" } ] } }```

**Comment by lewisgoddard** [comment]
image

File: _scripts/pages/store/index.js [hunk 1]
```diff
@@ -47,8 +47,32 @@ Promise.all([jQuery, modal]).then(([$]) => {
...
+         *
+         * @param {String} printfulId - the variant printful_id
+         *
+         * @return {Void}
+         */
+        var updateShippingEstimate = function (printfulId) {
+            var $m = $('.js-leanmodal-active .modal__shipping')
+            // GET /api/gelocate
+            // with the parameters "shipping" and "item" (printful variant id)
+            $.getJSON(baseUrl + 'api/geolocate?shipping&item=' + printfulId, function (data) {
```
Comment (btkostner) [inline _scripts/pages/store/index.js]:
```js
            $.getJSON(baseUrl + 'api/geolocate?shipping&item=' + printfulId, function (data) {
                // Update price information
				try {
				  const cost = data['shipping']['estimates'][0]['cost']

				  if (cost != null) {
                    $m.text('+ $' + cost + ' estimated shipping')
                  }
                } catch (e) {}
            })
```

I would use a catch block here instead. That way it does not fail if a nested value does not exist

File: api/geolocate.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    $result = array(
+        'ip' => $ip,
+        'shipping' => array(
+            'address' => $estimatedAddress,
+        ),
+    );
+
+    if ( !empty($_GET['item']) ) {
+        require_once __DIR__.'/../_backend/store/address.php';
+        require_once __DIR__.'/../_backend/store/api.php';
```
Comment (btkostner) [inline api/geolocate.php]: These have no side effects (I'm like 99.99% sure), so they should go up top on the page.

File: api/geolocate.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    $estimatedAddress = getCurrentLocation($ip);
+
+    $result = array(
+        'ip' => $ip,
+        'shipping' => array(
+            'address' => $estimatedAddress,
+        ),
+    );
+
+    if ( !empty($_GET['item']) ) {
```
Comment (btkostner) [inline api/geolocate.php]: Something is failing when giving an item ID. I have no checked the server for the error though.

Comment (lewisgoddard) [reply api/geolocate.php]: Could you take a look when you get the chance? I can't reproduce.

Comment (lewisgoddard) [reply api/geolocate.php]: I might have fixed this, local IPs can't geolocate.

**Comment by lewisgoddard** [comment]
image

**Comment by btkostner** [comment]
I would get a double check from @cassidyjames or @danrabbit but code works on staging

**Pull request merged**